File: perl/gol_bug.pl
#!/usr/bin/perl

=pod
Here, X --> alive cell
      O --> dead cell

Input Pattern : 
0 0 0 0 0 
0 0 X 0 0
0 0 X 0 0
0 0 X 0 0 
0 0 0 0 0
Output Pattern :
If the generation number is odd : | If the generation number is even :
								  |
0 0 0 0 0                         | 0 0 0 0 0
0 0 0 0 0                         | 0 0 X 0 0
0 X X X 0                         | 0 0 X 0 0
0 0 0 0 0                         | 0 0 X 0 0
0 0 0 0 0                         | 0 0 0 0 0
								  |
To print for debugging use :
print <variable>;  ......To print scalar quantities like int,string
print Dumper(\@arr);  .......To print data structure like arrays,hash
=cut

use Data::Dumper;
use warnings;
no warnings 'experimental::smartmatch';

my $i = 0;
# number of generations/iterations
my $gen = 5;
chomp($gen);
# input array
my @inp_arr = ([0,0,0,0,0],[0,0,1,0,0],[0,0,1,0,0],[0,0,1,0,0],[0,0,0,0,0]);
my @checker_arr = ([0,0,0,0,0],[0,0,0,0,0],[0,1,1,1,0],[0,0,0,0,0],[0,0,0,0,0]);
my @final_arr;


## helper
sub getValue {
    my($x, $y ) = @_;
    my $s = 1;
    my $xbound = scalar @inp_arr;
    my $ybound = scalar @inp_arr;
        if ($x < 0 || $y < 0 || $x > $xbound-$s || $y > $ybound-$s) {
        return 0;
        } else {
            return($inp_arr[$x][$y]);
        }
}

# test function
sub tester {
    my @list = @_;
    $checksum = $list[0];
    if ($checksum == 99) {
        if (@final_arr ~~ @checker_arr){
            print "Tests passed\n";
        } else {
            print "Wrong patterns generated\n"
        }
    } else {
        print "\nFailed!!\n";
    }
}

## main function
sub main {

my $check = 0;
@print_arr = map { [@$_] } @inp_arr;
for (my $i = 0; $i < $gen; $i++){
    #print "Generation: $i\n";
	#print map((map($_ ? "X " : "O ", @$_), "\n"), @inp_arr);
    #print "\n\n";
    my $row = @inp_arr;
    for my $j (0 .. $inp_arr) {
        if ($j == 2){
        $check = 99;
        }
        for $k (0..$#{ $inp_arr[$i] }) { 
            my $count = 0;
        for ( [$j-1, $k-1], [$j-1, $k], [$j-1, $k+1],[ $j, $k-1],[ $j, $k+1],[ $j+1, $k-1], [ $j+1, $k], [ $j+1, $k+1]) {
            my $aa = $_->[0];
            my $bb = $_->[1];
            if (getValue($aa,$bb) == 1){
            $count++;
            }

        }
        # checker to make cell dead/alive
        if ($inp_arr[$j][$k] == 0){
                if ($count == 2 || $count == 3) {
                    $print_arr[$j][$k] = 1;
                } else {
                    $print_arr[$j][$k] = 0;
		}
        }
        # checker to make cell dead/alive
        if ($inp_arr[$j][$k] == 1) {
                if ($count == 3) {
                    $print_arr[$j][$k] = 1;
                } else {
                    $print_arr[$j][$k] = 0;
        }
        }

	}
    }

    @inp_arr = map { [@$_] } @print_arr;
    }
@final_arr = map { [@$_] } @inp_arr;
tester($check)
}
main()

File: Go/gol_bug.go
/*

Input Pattern : 

0 0 0 0 0 
0 0 1 0 0
0 0 1 0 0
0 0 1 0 0 
0 0 0 0 0

Output Pattern :

If the generation number is odd : | If the generation number is even :
								  |
0 0 0 0 0                         | 0 0 0 0 0
0 0 0 0 0                         | 0 0 1 0 0
0 1 1 1 0                         | 0 0 1 0 0
0 0 0 0 0                         | 0 0 1 0 0
0 0 0 0 0                         | 0 0 0 0 0
								  |
To print for debugging use :

fmt.Print("String", variable)
fmt.Println("String", variable) - adds a newline at the end

*/

package main

import (
	"fmt"
)

//main function
func main() {
	
	//Initial pattern
	var input_pattern = [5][5]int{
		{0,0,0,0,0},
		{0,0,1,0,0},
		{0,0,1,0,0},
		{0,0,1,0,0},
		{0,0,0,0,0},
	}

	//Defining the number of generations
	var number_of_generations int = 3 

	//Loop which calls the generate_next_pattern function 
	for generation:=0; generation < number_of_generations; generation++ {

	var output_pattern [5][5]int= generate_next_pattern(input_pattern)
	input_pattern = output_pattern
	
	}

	var correct bool = test()

	if(correct == false){
		fmt.Println("Wrong patterns generated")
	} else {
		fmt.Println("Tests passed")
	}
}

//This function is where the next pattern is generated. Find bugs in this function to generate the correct next pattern.
func generate_next_pattern(input_pattern [5][5]int) [5][5]int {

	//initializing new_pattern array  which will be the next generation
	new_pattern := [5][5]int{
		{0,0,0,0,0},
		{0,0,0,0,0},
		{0,0,0,0,0},
		{0,0,0,0,0},
		{0,0,0,0,0}}
	
	//looping through every element in input_pattern
	for row:=0; row < input_pattern.length; row++ {
		for column:=0; column < input_pattern[row].length; column++ {
			
			var count int = 0

			//looping through neighbours to calculate sum
			for i:=(row-1); i<=(row+1); i++ {
				for j:= (column-1); j<=(column + 1); j++ {
					if (i>=0 && i<=4 && j>=0 && j<=4 && !(i==row && j==column)) {
						count = count + input_pattern[i][j]	
					} 
				}
			}
			
			if (input_pattern[row][column] == 0) {
				if (count==3 || count == 2) {
					new_pattern[row][column] = 1} else{
				}
			} else {
				if (count==3) {
					new_pattern[row][column] = 1} else{
						new_pattern[row][column] = 0
				}
			}
		}
	}

	return(new_pattern)
}	

//Test function
func test() bool {

	var cur_pattern_1 = [5][5]int{
		{0,0,0,0,0},
		{0,0,1,0,0},
		{0,0,1,0,0},
		{0,0,1,0,0},
		{0,0,0,0,0},
	}

	for i:= 1; i<=4; i++ {
		cur_pattern_1 = generate_next_pattern(cur_pattern_1)

		if(i%2 == 0){
		
			var correct_pattern_1 = [5][5]int{
				{0,0,0,0,0},
				{0,0,1,0,0},
				{0,0,1,0,0},
				{0,0,1,0,0},
				{0,0,0,0,0},
			}
	
	
			if(cur_pattern_1 == correct_pattern_1){
				continue
			} else{
				return(false)
			}
	
		} else{
	
			var correct_pattern_1 = [5][5]int{
				{0,0,0,0,0},
				{0,0,0,0,0},
				{0,1,1,1,0},
				{0,0,0,0,0},
				{0,0,0,0,0},
			}
	
			if(cur_pattern_1 == correct_pattern_1){
				continue
			} else{
				return(false)
			}
	
		}

	}

	var cur_pattern_2 = [5][5]int{
		{0,0,0,0,0},
		{0,1,1,0,0},
		{0,1,1,0,0},
		{0,0,0,0,0},
		{0,0,0,0,1},
	}

	for i:= 1; i<=2; i++ {
		cur_pattern_2 = generate_next_pattern(cur_pattern_2)

		
		
			var correct_pattern_1 = [5][5]int{
				{0,0,0,0,0},
				{0,0,1,0,0},
				{0,0,1,0,0},
				{0,0,1,0,0},
				{0,0,0,0,0},
			}
	
	
			if(cur_pattern_1 == correct_pattern_1){
				continue
			} else{
				return(false)
			}
	

	}

	return(true)
	
}